pkg/runtime/field_export_reconciler.go

```diff
@@ -0,0 +1,674 @@
+	}
+	secret.Data[key] = []byte(sourceValue)
+
+	ackrtlog.DebugFieldExport(r.log, desired, "patching target secret")
+	err = r.kc.Patch(ctx, secret, patch)
+	if err != nil {
+		return err
+	}
+	ackrtlog.InfoFieldExport(r.log, desired, "patched target secret")
+
+	return nil
+}
+
+func (r *fieldExportReconciler) FilterAllExports(
+	ctx context.Context,
+	gk metav1.GroupKind,
+	nsn types.NamespacedName,
+) ([]ackv1alpha1.FieldExport, error) {
+	listed := &ackv1alpha1.FieldExportList{}
+	opts := []client.ListOption{
+		client.InNamespace(nsn.Namespace),
+	}
+	if err := r.apiReader.List(ctx, listed, opts...); err != nil {
+		return []ackv1alpha1.FieldExport{}, err
+	}
+
+	exports := []ackv1alpha1.FieldExport{}
+	for _, export := range listed.Items {
+		// Ensure we are working with managed exports
+		if !k8sctrlutil.ContainsFinalizer(&export, fieldExportFinalizerString) {
+			continue
+		}
+
+		// Check the reference matches our source resource
+		if !strings.EqualFold(export.Spec.From.Resource.Kind, gk.Kind) ||
+			!strings.EqualFold(*export.Spec.From.Resource.Name, nsn.Name) {
+			continue
+		}
+
+		exports = append(exports, export)
+	}
+
+	return exports, nil
+}
+
+// onError will patch the FieldExport with the given error and return the
+// same error back
+func (r *fieldExportReconciler) onError(
+	ctx context.Context,
+	res *ackv1alpha1.FieldExport,
+	err error,
+) error {
+	r.patchRecoverableCondition(ctx, res, err)
+	return err
+}
+
+// onSuccess will patch the FieldExport with a synced condition and
+// return any errors that occurred while patching
+func (r *fieldExportReconciler) onSuccess(
+	ctx context.Context,
+	res *ackv1alpha1.FieldExport,
+) error {
+	r.clearConditions(ctx, res)
+	return nil
+}
+
+// clearConditions removes all conditions from the field export CR.
+func (r *fieldExportReconciler) clearConditions(
+	ctx context.Context,
+	res *ackv1alpha1.FieldExport,
+) error {
+	base := res.DeepCopy()
+	res.Status.Conditions = []*ackv1alpha1.Condition{}
+	return r.patchStatus(ctx, res, base)
+}
+
+// patchRecoverableCondition updates the recoverable condition status of the
+// field export CR. The resource passed in the parameter gets updated with the
+// conditions
+func (r *fieldExportReconciler) patchRecoverableCondition(
+	ctx context.Context,
+	res *ackv1alpha1.FieldExport,
+	err error,
+) error {
+	base := res.DeepCopy()
+
+	// Recoverable condition
+	var recoverableCondition *ackv1alpha1.Condition = nil
+	for _, condition := range res.Status.Conditions {
+		if condition.Type == ackv1alpha1.ConditionTypeRecoverable {
+			recoverableCondition = condition
+			break
+		}
+	}
+
+	if recoverableCondition == nil {
+		recoverableCondition = &ackv1alpha1.Condition{
+			Type: ackv1alpha1.ConditionTypeRecoverable,
+		}
+		res.Status.Conditions = append(res.Status.Conditions, recoverableCondition)
+	}
+
+	var errMessage string
+	if err != nil {
+		errMessage = err.Error()
+		recoverableCondition.Status = corev1.ConditionFalse
```

Comment:
If the error is not Terminal error then recoverable should be set to true, correct?

Reply:
Oh wow I had this logic backwards. If there is an error, set `Recoverable` to `true` and set the message to that error message. If there is not error, set `Recoverable` to `false` and remove any message.

Reply:
>If there is not error, set Recoverable to false and remove any message.

Why not clear the recoverable condition in this case?

Reply:
We won't ever reach the code path where `err == nil` in this reconciler anyway. We only call this method when we have an error. I'll just remove that branch so it's not confusing.